### ISSUE TITLE
Corrupted CONTINUE Cards

### ISSUE THREAD

**Issue by kadrlica** [esheldon/fitsio#130]
I've been using the stop-gap fix in `v0.9.12rc1` to treat `CONTINUE` keys like comments (#117 and #119). So far things have been working well (over ~100k images processed). However, I just hit this error:
```
ValueError: bad card 'CONTINUE= '        '           /   '&' / Current observing orogram'.  CONTINUE not supported
```

Below is a snippet of the offending header:
```
CAMSHUT = 'Open    '           / Camera shutter at exposure start               
PROGRAM = 'Catching Supernovae in the Act with KEGS (Kepler Extra-Galactic &'   
CONTINUE= '        '           /   '&' / Current observing orogram              
OBSERVER= 'DJJ     '           / Observer name(s)                               
```

Something is amiss here. The [FITS convention](https://fits.gsfc.nasa.gov/registry/continue_keyword.html) states that keywords should "have the name  CONTINUE in bytes 1 through 8, and have space characters in bytes 9 and 10" (i.e., no "="). The header above doesn't obey this convention, and thus can be considered corrupt.

Now, we could just claim this is a corrupt header and wash our hands of the matter. However, I'm pretty sure that this header got corrupted by being processed by an older version of `fitsio` that was not treating `CONTINUE` cards correctly. The original header before processing:
```
CAMSHUT = 'Open    '           / Camera shutter at exposure start
PROGRAM = 'Catching Supernovae in the Act with KEGS (Kepler Extra-Galactic &'
CONTINUE  'Survey) and DECam&'
CONTINUE  '&' / Current observing orogram
OBSERVER= 'DJJ     '           / Observer name(s)
```

This type of corruption is consistent with how older versions of `fitsio` dealt with `CONTINUE` cards. Even if the corrupted header does not strictly conform to the FITS convention, it would be nice if we could at least read it with newer versions of `fitsio`.

The change would be relatively minor, just adding `or front=='CONTINU'` [here](https://github.com/esheldon/fitsio/blob/cf0cfec03b9971eeb733a7e85ddb16ab0a0bb015/fitsio/fitslib.py#L4385). However, this would break with the standard convention, which states that CONTINUE will not be followed by "=".

Thoughts from @esheldon?

**Comment by esheldon** [comment]
Actually it should have been read as a normal key, but we still have this in place

https://github.com/esheldon/fitsio/blob/cf0cfec03b9971eeb733a7e85ddb16ab0a0bb015/fitsio/fitslib.py#L4432

There is no need for that there, unless I'm missing something.

**Comment by kadrlica** [comment]
I think you are right, we could remove that check and  `CONTINUE=` would be read as a normal key.  I left that check in there because I thought it would be help us catch broken headers (which it did), but I think you are right that by removing it we could read these corrupted headers and still be consistent with the FITS convention.

Want me to submit a PR?

**Comment by esheldon** [comment]
yes, please do, and can you add a unit test as well?

**Comment by kadrlica** [comment]
The read seems ok, but there is an issue that keys named `CONTINUE` are not written normally:
https://github.com/esheldon/fitsio/blob/f6482556354506b632ce8d1377ba84382db3b2b5/fitsio/fitslib.py#L1404-L1405

So we could add some logic here to distinguish `CONTINUE=` from `CONTINUE`, or we could parse `CONTINUE=` like a real `CONTINUE` card from the start.

**Comment by kadrlica** [comment]
This decision may be out of our hands, since `cfitsio` sets the keyclass of cards that start with `CONTINUE` to `TYP_CONT_KEY` regardless of whether or not they have an `'='`:
https://github.com/esheldon/fitsio/blob/f6482556354506b632ce8d1377ba84382db3b2b5/cfitsio3370patch/fitscore.c#L8099-L8100

In light of this, I'm re-proposing the change in 1569a44 to treat all cards that start with `CONTINU` as `CONTINUE` keys regardless of whether they have a space. The downside is that it will overwrite `CONTINUE= ` with `CONTINUE ` when a header is read and written (this is the behavior for `COMMENT =` and `HISTORY =` corrupted keys as well).

To make things more complicated, these three keys will evaluate to different values on a read and write:
```
CONTINUE  'key'   -> CONTINUE  'key' # two spaces in 9-10 place (fits convention)
CONTINUE= 'key'  -> CONTINUE 'key' # one space in 9 (not convention)
CONTINUE = 'key' -> CONTINUE = 'key # '=' in 10 with spaces in 9 & 11
```
I'm not sure exactly what is going on in the second one, but I'm guessing it is stripping the `'='` and not padding out the value.

**Comment by esheldon** [comment]
Where do we stand on this?

**Comment by kadrlica** [comment]
I hacked my personal version so that I can read the broken continue keys, but I haven't done anything to actually fix the problem.

**Comment by rmjarvis** [comment]
@kadrlica Ha.  I should have looked for an existing issue about this before making my PR (#133).  It looks like we both have person versions with basically the same fix.  :)

I didn't completely follow all the details about how this breaks FITS convention, but I think the priority should be to allow reading these non-compliant files and not worry so much about whether re-writing them results in a slightly different string value.

In fact I think a warning such as the following would be sufficient:
```
if self.has_equals():
    import warnings
    mess = "Found key %s with an = sign. "%card_string[:8]
    mess += "This is not compliant with FITS standards. "
    mess += "There may be slight inconsistencies if you write this back out to a file."
    warnings.warn(mess, FITSRuntimeWarning)
```
I can add that to my PR if you think this is OK.

**Comment by kadrlica** [comment]
Sounds good to me. We both implemented the same temporary fix, so I'm perfectly happy to have yours committed. I think that we should leave this issue open.

**Comment by kadrlica** [comment]
@esheldon @rmjarvis I think that #133 may have generated some new momentum. As I see it, the core issue is whether we are ok with `fitisio` implicitly correcting corrupted `CONTINUE` keys when a file is read and written *without* explicit alteration by the user.

If we are ok with `fitsio` changing `CONTINUE` keys under the hood, then I think we have a path forward. Otherwise, we are going to need to write some code to preserve corrupted `CONTINUE` keys.

Thoughts?

**Comment by esheldon** [comment]
This sort of issue has come up a few times in different contexts.

One example is with regard to string columns in binary tables.  In `cfitsio` strings are always space padded out to the declared field size when writing.  This means you don't get out what you put in.

This has caused some consternation among users.  And it is also not consistent with the FITS standard, which says the string should be null padded (`cfitsio` and the fortran version predate the standard).  It also means files written by pyfits (which is standards compliant in this case) can't be exactly copied to new files when using fitsio for the reading/writing.

We should consider things like this when making a decision.

**Comment by kadrlica** [comment]
Yep, these decisions are what stalled progress on this issue before.

**Comment by esheldon** [comment]
@kadrlica fitsio master now supports null padded strings, so we have already broken what people might expect from cfitsio :)  Do you want to try a PR again to address the remaining issues, those not addressed by Mike's PR?  we are bout to release v1.0.0 so this is probably a good time

**Comment by kadrlica** [comment]
Ok, I can put it on my todo list. What is the timescale?

**Comment by esheldon** [comment]
I think it is important to have this one fixed for the v1 release.  Could you get to it on a weekish timescale?

**Comment by kadrlica** [comment]
I think I can give it a shot.

**Comment by esheldon** [comment]
May I suggest that we go ahead and merge #217 and  perhaps think about writing a stand alone tool to fix headers corrupted by old fitsio?

**Comment by kadrlica** [comment]
I agree that #217 should be decoupled from this issue. I don't think a standalone tool is necessary, unless it is requested by someone else. I think this issue can probably be closed once #217 is merged.

**Comment by esheldon** [comment]
I think a completely new parser might be needed in order to reconstruct the contents of these corrupted continue.
